### PR TITLE
CM-296 When to Use Static Classes in C#

### PR DESCRIPTION
**Pull request by martinchege324**

### CODE REVIEW

File: csharp-advanced-topics/StaticClasses/StaticClassesTest/StaticClassUnitTest.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        [TestMethod]
+        public void GivenStudentDate_ThenReturnCorrectAge()
+        {
+            //assign variables to the Student class
+            Program.Student.Id = 1;
+            Program.Student.Name = "John Doe";
+            Program.Student.DateOfBirth = new DateTime(1994, 12, 31);
+            int expectedAge = 26;
+            int actualAge = Program.Student.CalculateAge(Program.Student.DateOfBirth);
+            Assert.AreEqual(expectedAge, actualAge);
```
Comment (CodeMazeBlog) [inline csharp-advanced-topics/StaticClasses/StaticClassesTest/StaticClassUnitTest.cs:19]: Use newlines to separate the Arrange/Act/Assert parts of the tests.

**Review by CodeMazeBlog — COMMENTED** [review]
Small changes, mainly formatting, and some naming.

EDIT: I've found three different PRs on this topic. You've probably created a new one each time to make the build pass. If you do that please close the old ones. 
The other thing is naming the PRs. Please use the naming convention we use, which is to include the issue id and name in the PR title.
Also, a good way to go around creating PRs is to create a new branch on your fork and then to create a pull request from that branch. I see you had some leftover code from the evaluation article and that you needed to do some cleaning. By creating a new branch, you can avoid that and the PR will contain just the commits you've made for the particular article you're working on. Please avoid merging everything from the main branch in the future.

**Comment by martinchege324** [comment]
Thanks for the feedback, I have made those changes.

File: csharp-advanced-topics/StaticClasses/StaticClasses/CollegeStudent.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        //non-static methods
+        //function to return the student's age
+        public int CalculateAge(DateTime DateOfBirth)
+        {
+            //get today's date 
+            var _today = DateTime.Today;
+
+            //calculate age
+            var _age =_today.Year - DateOfBirth.Year;
```
Comment (CodeMazeBlog) [inline csharp-advanced-topics/StaticClasses/StaticClasses/CollegeStudent.cs:27]: Local variables shouldn't have underscores. Only the private fields should have those, like the _name. Please go through the entire solution and fix it where applicable.

File: csharp-advanced-topics/StaticClasses/StaticClasses/Program.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        //non-static class student implementation 
+        //the class has to be instantiated
+        CollegeStudent student = new CollegeStudent();
+        student.Id = 1;
+        student.Name = "John Doe";
+        student.DateOfBirth = new DateTime(1994, 12, 31);
+        Console.WriteLine("The student details using a non-static class: " + student.StudentDetails());
+    }
+   
+    public static class Student
```
Comment (CodeMazeBlog) [inline csharp-advanced-topics/StaticClasses/StaticClasses/Program.cs:23]: Should go to the separate file.

File: csharp-advanced-topics/StaticClasses/StaticClasses/Program.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        //static class implementation
+        //no instantiation needed
+        Student.Id = 1;
+        Student.Name = "John Doe";
+        Student.DateOfBirth = new DateTime(1994, 12, 31);
+        Console.WriteLine("The student using a static class: " + Student.StudentDetails());
+
+        //non-static class student implementation 
+        //the class has to be instantiated
+        CollegeStudent student = new CollegeStudent();
```
Comment (CodeMazeBlog) [inline csharp-advanced-topics/StaticClasses/StaticClasses/Program.cs:16]: You can improve this by using var, or new(). We prefer using var, but it's up to you.

File: csharp-advanced-topics/StaticClasses/StaticClassesTest/StaticClassUnitTest.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+            Assert.AreEqual(expectedAge, actualAge);
+        }
+
+        [TestMethod]
+        //test the non-static version of the same class
+        public void GivenStudentAgeNonStatic_ThenReturnCorrectAge() 
+        {
+            //instantiate non static class
+            CollegeStudent student = new CollegeStudent();
```
Comment (CodeMazeBlog) [inline csharp-advanced-topics/StaticClasses/StaticClassesTest/StaticClassUnitTest.cs:28]: You can improve this by using var, or new(). We prefer using var, but it's up to you.

File: csharp-advanced-topics/StaticClasses/StaticClassesTest/StaticClassUnitTest.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        //test the type of the object returned by the static class is not equal to the non-static class type
+        public void GivenStaticAndNonStaticClass_ThenReturnConfirmNotSame() 
+        {
+            Program.Student.Id = 1;
+            Program.Student.Name = "John Doe";
+            Program.Student.DateOfBirth = new DateTime(1994, 12, 31);
+            var staticVersion = Program.Student.StudentDetails;
+
+            //lets compare whether the class is the same as it's non-static counterpart
+            CollegeStudent student = new CollegeStudent();
```
Comment (CodeMazeBlog) [inline csharp-advanced-topics/StaticClasses/StaticClassesTest/StaticClassUnitTest.cs:48]: You can improve this by using var, or new(). We prefer using var, but it's up to you.

**Comment by martinchege324** [comment]
Thanks for the feedback, I have made those changes thanks.

**Comment by CodeMazeBlog** [comment]
Everything looks alright now.
Please create a separate branch for the next article before you create a PR so we can have a clean commit history.

**Pull request merged**